feat: Introduce a separate parsing_method rule

rpl opened the pull request:
This PR move shared bits of the `method` rule into a `base_method.js` module and then reuse the helper function exported by that method to create two separate `method` and `parsing_method` rules.

The purpose of splitting the `method` rule into two separate rules is to report violations detected by the `parsing_method` rule as warnings instead of errors by default, and allow developers to opt-in/opt-out on parsing methods violations separately from the `method` rule.

mozfreddyb commented:
Right. This would help with #81, which is a long-standing issue for some users.

mozfreddyb commented on docs/rules/parsing_method.md:
```diff
@@ -0,0 +1,12 @@
+# parsing_method
+
+The _parsing_method_ rule in _eslint-plugin-no-unsanitized_ perform basic security
+checks for function calls that parse strings into new Document instances. The idea
```
```suggestion
checks for function calls that parse strings into new Document or DocumentFragment instances. The idea
```

mozfreddyb commented on docs/rules/parsing_method.md:
```diff
@@ -0,0 +1,12 @@
+
+The _parsing_method_ rule in _eslint-plugin-no-unsanitized_ perform basic security
+checks for function calls that parse strings into new Document instances. The idea
+of these checks is to allow developers to opt-in/opt-out detecting use of these methods,
```
```suggestion
of these checks is to allow developers to opt-in/opt-out of detecting use of these methods,
```

rpl commented on tests/rules/parsing_method.js:
```diff
@@ -0,0 +1,49 @@
+
+        {
+            code: "var doc = Document.parseHTMLUnsafe(badness);",
+            errors: [
```
> is there a way to make it a warning by moving "parsing" stuff into its own specific check? What happens if people would upgrade to a new linter version that includes "parsing_method"? Would they get new breakage? Would they just silently not get the new rule because of their settings?
> 
> If this breaks people, I'd be slightly more inclined to move the whole parsing_method into a completely different repo (that I will happily support)...

@mozfreddyb This line in the module declaring the plugin rules https://github.com/mozilla/eslint-plugin-no-unsanitized/pull/244/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R22 should be the one setting the level of the issues identified by the parsing_method rule to warning instead of error, but I'll double-check by trying that locally to confirm that is actually the case.

mozfreddyb replied:
Oh I must have skipped that. No, that should totally work (and hopefully be configurable by users). Nice!

Rob--W commented:
Is the purpose of this to add validation of `Document.parseHTMLUnsafe`?

Don't forget to also add support for `setHTMLUnsafe`: https://developer.mozilla.org/en-US/docs/Web/API/Element/setHTMLUnsafe

rpl commented:
> Is the purpose of this to add validation of `Document.parseHTMLUnsafe`?
> 
> Don't forget to also add support for `setHTMLUnsafe`: https://developer.mozilla.org/en-US/docs/Web/API/Element/setHTMLUnsafe

@Rob--W That is already covered by the 'method' rule, see https://github.com/mozilla/eslint-plugin-no-unsanitized/blob/7835bfdd6f48e8e4f49ade7efd59318acc49e42b/lib/rules/method.js#L43

rpl commented on index.js:
```diff
@@ -11,13 +11,15 @@ const plugin = {
 const rules = {
     "no-unsanitized/property": "error",
     "no-unsanitized/method": "error",
+    "no-unsanitized/parsing_method": "warn",
```
@mozfreddyb I'm glad that I did actually give it an explicit try locally, this was meant to be "warn" instead of "warning". It wasn't covered by any test and so easy to miss, I thought to take a look into adding at least a very basic integration test (one that would have caught the issue in the PR and before releasing a new version).

We can also decide to move that proposed addition to the tests and CI job in a separate PR if we prefer.

mozfreddyb replied:
I think it would be good to have tests in the same moment that we add the functionality. :)

rpl replied:
100%, I couldn't agree more

rpl commented:
@mozfreddyb I opened up this PR for a review pass on your side (in the meantime I applied the fixes for the typos in the markdown file, fixed the recommended config, and the README.md file which had a syntax error, and added a base integration test executed by the github workflow, which would have caught the issue with the typo in the severity level of the new rule in the previous version of this PR).

mozfreddyb commented on lib/rules/parsing_method.js:
```diff
@@ -0,0 +1,40 @@
+
+    // check first parameter to `Document.parseHTMLUnsafe()`
+    parseHTMLUnsafe: {
+        objectMatches: ["Document"],
```
```suggestion
        objectMatches: ["document"],
```

Our matching is case-insensitive and use lowercase `document` elsewhere in the code.

mozfreddyb commented on tests/rules/parsing_method.js:
```diff
@@ -0,0 +1,62 @@
+
+    valid: [
+        {
+            code: "var doc = DOMParser.parseFromString('static string');",
```
That's not entirely how the function is used. People create a `DOMParser()` instance and then call `parseFromString` on it. https://developer.mozilla.org/en-US/docs/Web/API/DOMParser/parseFromString#examples

rpl replied:
/me feeling stupid, I even knew that, I was too focused on the other one and made the mistake

mozfreddyb commented on tests/rules/parsing_method.js:
```diff
@@ -0,0 +1,62 @@
+         */
+
+        {
+            code: "var doc = DOMParser.parseFromString(badness);",
```
Same as above. `parseFromString` is not a static method.

mozfreddyb reviewed (changes requested):
Looks great. Just a few things missing :)

rpl commented on tests/rules/parsing_method.js:
```diff
@@ -0,0 +1,62 @@
+            errors: [
+                {
+                    message:
+                        /Unsafe call to DOMParser.parseFromString for argument 0/,
```
That feels like a good point to be too, let me take a look to what additional change may be needed in this PR if we would like to use slightly different message for the separate parsing_method rule.

rpl replied:
@mozfreddyb I have included in [afe2df6](https://github.com/mozilla/eslint-plugin-no-unsanitized/pull/244/commits/afe2df60e3e1c7920b7b6e40967104601d227a88) one possible approach to customize the message (for now I went for the simpler "Potentially unsafe call..." one, but if the approach looks reasonable we can tweak the message further).

I have also changed the objectMatches for the DOMParser parseFromString method to 'parser' for now.

I was actually tempted to go even for just 'parse' or 'pars' (so that it would also match other variations, like 'htmlParsing' or 'parseHTML' or 'parseUtils'), wdyt?

rpl commented on tests/integration/test-eslint-flat-config.js:
```diff
@@ -0,0 +1,81 @@
+
+describe("eslint-flat-config", function () {
+    before(function () {
+        // Only run these integration tests while running on esling
```
:facepalm:

rpl commented on tests/rules/parsing_method.js:
```diff
@@ -0,0 +1,62 @@
+
+    // Examples of code that should trigger the rule
+    invalid: [
+        /* XXX Do NOT change the error strings below without review from freddy:
+         * The strings are optimized for SEO and understandability.
+         * The developer can search for them and will find this MDN article:
+         *  https://developer.mozilla.org/en-US/Firefox_OS/Security/Security_Automation
+         */
```
eh, the start of the inline comment felt liike something to keep in the new test file and I missed to notice the historical developer.mozilla.org related to Firefox OS being there

mozfreddyb commented on tests/rules/parsing_method.js:
```diff
@@ -0,0 +1,81 @@
+    // Examples of code that should not trigger the rule
+    // XXX this does not find z['innerHTML'] and the like.
+
+    valid: [
+        {
+            code: "var doc = parser.parseFromString('static string');",
+        },
+        {
+            code: "var doc = Document.parseHTMLUnsafe('static string');",
+        },
+    ],
```
What would a sanitized version of this look like?

`parseHTMLUnsafe(sanitize(badness))` or would `sanitizeDocument(parseHTMLUnsafe(badness))` be more likely? I am leaning to (1) and would like to suggest to add a simplified test case here

mozfreddyb commented on docs/rules/parsing_method.md:
```diff
@@ -0,0 +1,13 @@
+# parsing_method
+
+The _parsing_method_ rule in _eslint-plugin-no-unsanitized_ performs basic security
+checks for function calls that parse strings into new Document or DocumentFragment
+instances. The idea of these checks is to allow developers to opt-in/opt-out of detecting
+use of these methods, separately from the `method` rule which is reporting violation
+as errors by default.
```
```suggestion
The _parsing_method_ rule in _eslint-plugin-no-unsanitized_ performs basic security
checks for function calls that parse strings into new Document or DocumentFragment
instances. It is close to impossible to automatically check whether the resulting
Document is actually used insecurely, so this offers a control at the parse step.
Therefore, these checks allow developers to opt-in/opt-out of detecting
use of these methods, separately from the `method` and `property` rule which is
reporting violation as errors by default.
```

mozfreddyb reviewed (commented):
Almost ready to go :) Thank you, Luca!
